HOTFIX: Fix lgtm.com alerts (dead code and out-of-bounds error)

zbazztian opened the pull request:
This fixes two alerts flagged on [lgtm.com](https://lgtm.com) for Apache Kafka.

1. This [dead code alert](https://lgtm.com/projects/g/apache/kafka/snapshot/6abe9363dbc8d7295055865233071ef51dce1f3a/files/tools/src/main/java/org/apache/kafka/trogdor/rest/RestExceptionMapper.java?sort=name&dir=ASC&mode=heatmap&excluded=false#xbfc4f8e26024a8f4:1) where `InvalidTypeIdException` indirectly extends `JsonMappingException`. The flagged condition with the type test appears after the type test for the latter and thus makes its body dead. I opted to change the order of the tests. Please let me know if this is the intended behavior.

2. The second commit addresses this [out-of-bounds alert](https://lgtm.com/projects/g/apache/kafka/snapshot/6abe9363dbc8d7295055865233071ef51dce1f3a/files/clients/src/main/java/org/apache/kafka/common/security/kerberos/KerberosRule.java?sort=name&dir=ASC&mode=heatmap&excluded=false#x40b686b508503559:1).

More alerts can be found [here](https://lgtm.com/projects/g/apache/kafka/alerts/). Note that my colleague Aditya Sharad addressed some of those in the now outdated #2939.

@mjsax I don't know any committers here, apart from you, so maybe you could have a look? How are you doing by the way? Long time no see...

Commit:
Fix out-of-bounds error.

mjsax reviewed (approved):
Thanks for the patch @sebastiansemmle ! I am not familiar with this part of the code, but the change seems reasonable to be. (I am not a committer and can't merge it though.)

LGTM.

Call for second review and merging @rajinisivaram 

@sebastiansemmle Doing good. Let's catch up via a non-public channel :)

rajinisivaram commented:
@sebastiansemmle Thanks for the PR. Changes look good. Would it possible to add unit tests for these changes? Thank you!

Commit:
Update documentation for KerberosRule.replaceParameters().

Commit:
Add unit test for KerberosRules.

rajinisivaram commented:
@sebastiansemmle The change to `RestExceptionMapper` was done under another PR along with unit tests and that has been committed. Can you update this PR to just contain the change to `KerberosRule` and perhaps add a test for that? Thank you!

zbazztian commented:
@rajinisivaram Thanks for reviewing this. I rebased and added the unit tests. I also updated the documentation for `KerberosRule.replaceParameters()` as it was contradicting the implementation. Let me know what you think.

Commit:
Remove tabs.

zbazztian commented:
Tabs removed.

rajinisivaram reviewed (approved):
@sebastiansemmle Thanks for the updates, LGTM. Will merge after PR builds complete.

zbazztian commented:
Thanks @rajinisivaram . The tests for JDK 8 failed, unfortunately. Not sure what the reason is though. It looks like some of the tests were skipped. Looks unrelated to my changes at first sight...

rajinisivaram commented:
retest this please

rajinisivaram commented:
@sebastiansemmle Yes, the failure looks unrelated. Re-running the tests anyway.

ijuma commented:
Retest this please

rajinisivaram commented:
retest this please

zbazztian commented:
retest this please

guozhangwang commented:
retest this please

mjsax commented:
@sebastiansemmle Can you rebase this PR? We merged a couple of fixed in `trunk` to stabilize the tests.

mjsax commented:
Retest this please.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/269/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/287/

mjsax commented:
Retest this please.

guozhangwang commented:
retest this please

mjsax commented:
Retest this please.

guozhangwang commented:
Merged to trunk, thanks @sebastiansemmle
